File: frontend/packages/studio/stores/bot-detail/src/types/agent.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
import { type PartialRequired } from '@coze-arch/bot-typings/common';
import type { DraftBotApi } from '@coze-arch/bot-api/playground_api';
import type {
  AgentInfo,
  ModelInfo,
  JumpConfig,
  MultiAgentSessionType,
} from '@coze-arch/bot-api/developer_api';
import type {
  LineType,
  WorkflowEdgeJSON,
} from '@flowgram-adapter/free-layout-editor';

import type { BotDetailSkill, BotSuggestionConfig } from './skill';
import type { RequiredBotPrompt } from './persona';

/** multi agent 相关数据 */
export interface BotMultiAgent {
  agents: Agent[];
  edges: WorkflowEdgeJSON[];
  connector_type: LineType;
  /** 用于保存 bot 类型节点的 bot 信息 */
  botAgentInfos: DraftBotVo[];
  /**
   * 会话接管方式配置
   * 默认为 flow 模式
   */
  chatModeConfig: ChatModeConfig;
}
/** 业务用到的 */
export interface AgentBizInfo {
  focused?: boolean;
}

export type Agent = PartialRequired<Omit<AgentInfo, 'work_info'>, 'id'> & {
  prompt: string;
  model: ModelInfo;
  skills: Pick<
    BotDetailSkill,
    'pluginApis' | 'workflows' | 'knowledge' | 'devHooks'
  >;
  system_info_all: Array<RequiredBotPrompt>;
  bizInfo: AgentBizInfo;
  jump_config: JumpConfig;
  suggestion: BotSuggestionConfig;
};

/** api 返回的 bot 信息中，部分字段是 json，本类型是 parse 后的类型 */
export type DraftBotVo = Omit<DraftBotApi, 'work_info'> & {
  work_info: {
    suggest_reply: BotSuggestionConfig;
  };
};

export type ChatModeConfig =
  | {
      /** 会话接管方式 */
      type: MultiAgentSessionType.Flow;
    }
  | {
      /** 会话接管方式 */
      type: MultiAgentSessionType.Host;
      /** 当前的 host 节点 id */
      currentHostId: string;
    };

export interface MultiSheetViewOpenState {
  left: boolean;
  right: boolean;
}

File: backend/types/errno/agent.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package errno

import (
	"github.com/coze-dev/coze-studio/backend/pkg/errorx/code"
)

// single agent: 100 000 000 ~ 100 999 999
const (
	ErrAgentInvalidParamCode               = 100000000
	ErrAgentSupportedChatModelProtocol     = 100000001
	ErrAgentResourceNotFound               = 100000002
	ErrAgentPermissionCode                 = 100000003
	ErrAgentIDGenFailCode                  = 100000004
	ErrAgentCreateDraftCode                = 100000005
	ErrAgentGetCode                        = 100000006
	ErrAgentUpdateCode                     = 100000007
	ErrAgentSetDraftBotDisplayInfo         = 100000008
	ErrAgentGetDraftBotDisplayInfoNotFound = 100000009
	ErrAgentPublishSingleAgentCode         = 100000010
	ErrAgentAlreadyBindDatabaseCode        = 100000011
	ErrAgentExecuteErrCode                 = 100000012
)

func init() {
	code.Register(
		ErrAgentPublishSingleAgentCode,
		"publish single agent failed",
		code.WithAffectStability(true),
	)

	code.Register(
		ErrAgentGetDraftBotDisplayInfoNotFound,
		"get draft bot display info failed",
		code.WithAffectStability(true),
	)

	code.Register(
		ErrAgentSetDraftBotDisplayInfo,
		"set draft bot display info failed",
		code.WithAffectStability(true),
	)

	code.Register(
		ErrAgentUpdateCode,
		"update single agent failed",
		code.WithAffectStability(true),
	)

	code.Register(
		ErrAgentGetCode,
		"get single agent failed",
		code.WithAffectStability(true),
	)

	code.Register(
		ErrAgentCreateDraftCode,
		"create single agent failed",
		code.WithAffectStability(true),
	)
	code.Register(
		ErrAgentIDGenFailCode,
		"gen id failed : {msg}",
		code.WithAffectStability(true),
	)

	code.Register(
		ErrAgentPermissionCode,
		"unauthorized access : {msg}",
		code.WithAffectStability(false),
	)

	code.Register(
		ErrAgentResourceNotFound,
		"{type} not found: {id}",
		code.WithAffectStability(false),
	)

	code.Register(
		ErrAgentSupportedChatModelProtocol,
		"unsupported chat model protocol : {protocol}",
		code.WithAffectStability(false),
	)

	code.Register(
		ErrAgentInvalidParamCode,
		"invalid parameter : {msg}",
		code.WithAffectStability(false),
	)

	code.Register(
		ErrAgentAlreadyBindDatabaseCode,
		"already bind database : {msg}",
		code.WithAffectStability(false),
	)
	code.Register(
		ErrAgentExecuteErrCode,
		"systemt error",
		code.WithAffectStability(true),
	)
}
